app/src/main/java/org/p2p/wallet/user/repository/UserInMemoryRepository.kt

```diff
@@ -3,60 +3,93 @@ package org.p2p.wallet.user.repository
         pricesFlow.value = prices
     }
 
     override fun getPriceByToken(symbol: String): TokenPrice? =
         pricesFlow.value.firstOrNull { it.tokenSymbol == symbol }
 
-    private val decimalsFlow = MutableStateFlow<List<TokenData>>(emptyList())
+    private val allTokensFlow = MutableStateFlow<List<TokenData>>(emptyList())
 
     override fun setTokenData(data: List<TokenData>) {
-        decimalsFlow.value = data.sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.name })
+        allTokensFlow.value = data.sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.name })
     }
 
     override fun fetchTokens(searchText: String, count: Int, refresh: Boolean) {
         if (refresh) {
-            tokensFlow.value = emptyList()
+            searchTextByTokens.clear()
+            tokensSearchResultFlow.value = TokenListData()
         }
-        if (tokensFlow.value.size >= decimalsFlow.value.size) {
+        val allInMemoryTokens = allTokensFlow.value
+        val currentSearchTokenResultSize = tokensSearchResultFlow.value.size
+        if (currentSearchTokenResultSize >= allInMemoryTokens.size) {
             return
         }
-        val items = decimalsFlow.value
-        val offset = tokensFlow.value.size
 
-        if (searchText.isNotEmpty()) {
-            val result = items.filter {
-                searchText == it.symbol ||
-                    it.name.startsWith(searchText, ignoreCase = true)
+        when {
+            searchText.isEmpty() -> {
+                val defaultTokens = searchTextByTokens.getOrDefault(DEFAULT_TOKEN_KEY, emptyList())
+                val newSize = (defaultTokens.size - getPopularDecimals().size) + count
+                val needToLoadMore = newSize > defaultTokens.size
+                if (needToLoadMore) {
+                    searchTextByTokens[DEFAULT_TOKEN_KEY] = getPopularDecimals() + allInMemoryTokens.take(newSize)
+                }
+                setResult(DEFAULT_TOKEN_KEY)
+            }
+            searchText in searchTextByTokens -> {
+                val tokensBySearchText = searchTextByTokens.getOrDefault(searchText, emptyList())
+                val newSize = tokensBySearchText.size + count
+                val needToLoadMore = newSize > tokensBySearchText.size
+                if (needToLoadMore) {
+                    loadMoreTokensBySearchText(searchText, newSize)
+                }
+                setResult(searchText)
+            }
+            searchText !in searchTextByTokens -> {
+                loadMoreTokensBySearchText(searchText, count)
+                setResult(searchText)
             }
-            val indexes = findEdges(items = result, count = count, offset = offset)
-            appendTokens(result.subList(indexes.first, indexes.second))
         }
+    }
 
-        val indexes = findEdges(items = items, count = count, offset = offset)
-        val result = items.subList(indexes.first, indexes.second)
+    private fun setResult(key: String) {
```

Comment:
so, are you setting here only DEFAULT result? or search result?
i can see you are using only DEFAULT_TOKEN_KEY